domain/network/state/subnet.go

```diff
@@ -0,0 +1,447 @@
+	if spaceUUID == "" {
+		spaceUUIDValue = nil
+	}
+
+	insertSubnetStmt := `
+INSERT INTO subnet (uuid, cidr, vlan_tag, space_uuid, subnet_type_id)
+VALUES (?, ?, ?, ?, ?)`
+	insertSubnetAssociationStmt := `
+INSERT INTO subnet_association (subject_subnet_uuid, associated_subnet_uuid, association_type_id)
+VALUES (?, ?, 0)` // For the moment the only allowed association is 'overlay_of' and therefore its ID is hard-coded here.
+	retrieveUnderlaySubnetUUIDStmt := `
+SELECT uuid
+FROM   subnet
+WHERE  cidr = ?`
+	insertSubnetProviderIDStmt := `
+INSERT INTO provider_subnet (provider_id, subnet_uuid)
+VALUES (?, ?)`
+	insertSubnetProviderNetworkIDStmt := `
+INSERT INTO provider_network (uuid, provider_network_id)
+VALUES (?, ?)`
+	insertSubnetProviderNetworkSubnetStmt := `
+INSERT INTO provider_network_subnet (provider_network_uuid, subnet_uuid)
+VALUES (?, ?)`
+	insertAvailabilityZoneStmt := `
+INSERT INTO availability_zone (uuid, name)
+VALUES (?, ?)`
+	insertAvailabilityZoneSubnetStmt := `
+INSERT INTO availability_zone_subnet (availability_zone_uuid, subnet_uuid)
+VALUES (?, ?)`
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		// Add the subnet entity.
+		if _, err := tx.ExecContext(
+			ctx,
+			insertSubnetStmt,
+			uuid.String(),
+			cidr,
+			VLANTag,
+			spaceUUIDValue,
+			subnetType,
+		); err != nil {
+			return errors.Trace(err)
+		}
+
+		if subnetType == fanOverlaySegment {
+			// Retrieve the underlay subnet uuid.
+			var underlaySubnetUUID string
+			row := tx.QueryRowContext(ctx, retrieveUnderlaySubnetUUIDStmt, fanInfo.FanLocalUnderlay)
+			if err := row.Scan(&underlaySubnetUUID); err != nil {
+				return errors.Trace(err)
+			}
+			// Add the association of the underlay and the newly
+			// created subnet to the associations table.
+			if _, err := tx.ExecContext(
+				ctx,
+				insertSubnetAssociationStmt,
+				uuid.String(),
+				underlaySubnetUUID,
+			); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		// Add the subnet uuid to the provider ids table.
+		if _, err := tx.ExecContext(
+			ctx,
+			insertSubnetProviderIDStmt,
+			providerID,
+			uuid.String(),
+		); err != nil {
+			return errors.Trace(err)
+		}
+		// Add the subnet and provider network uuids to the
+		// provider_network_subnet table.
+		pnUUID, err := utils.NewUUID()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if _, err := tx.ExecContext(
+			ctx,
+			insertSubnetProviderNetworkIDStmt,
+			pnUUID.String(),
+			providerNetworkID,
+		); err != nil {
+			return errors.Trace(err)
+		}
+		// Insert the providerNetworkUUID into provider network to
+		// subnets mapping table.
+		if _, err := tx.ExecContext(
+			ctx,
+			insertSubnetProviderNetworkSubnetStmt,
+			pnUUID.String(),
+			uuid.String(),
+		); err != nil {
+			return errors.Trace(err)
+		}
+		for _, az := range availabilityZones {
+			// Add the availability zone.
+			azUUID, err := utils.NewUUID()
+			if err != nil {
+				return errors.Trace(err)
+			}
```

Comment:
This is a bit sad. I wonder if we can lift this out.

Reply:
I don't have a solution for this for now